src/PotentialTerm_OpenMM.cpp

```diff
@@ -0,0 +1,215 @@
+  cut_(0.8),         // in nm, Amber default
+  n_removed_dof_(0),
+  shakeH_(false),
+  shakeHeavy_(false)
+{}
+
+/// DESTRUCTOR
+PotentialTerm_OpenMM::~PotentialTerm_OpenMM() {
+# ifdef HAS_OPENMM
+  if (system_ != 0) delete system_;
+  if (context_ != 0) delete context_;
+# endif
+}
+
+#ifdef HAS_OPENMM
+
+/** This performs the actual openMM setup. */
+int PotentialTerm_OpenMM::OpenMM_setup(Topology const& topIn, Box const& boxIn,
+                                       CharMask const& maskIn, EnergyArray& earrayIn)
+{
+  using namespace Cpptraj::OMM;
+  mprintf("\tSetting up OpenMM.\n");
+  system_ = new OpenMM::System();
+  OpenMM::NonbondedForce* nonbond = new OpenMM::NonbondedForce();
+  system_->addForce( nonbond );
+  OpenMM::HarmonicBondForce* bondStretch = new OpenMM::HarmonicBondForce();
+  system_->addForce( bondStretch );
+  OpenMM::HarmonicAngleForce* angleStretch = new OpenMM::HarmonicAngleForce();
+  system_->addForce( angleStretch );
+  OpenMM::PeriodicTorsionForce* ptorsion = new OpenMM::PeriodicTorsionForce();
+  system_->addForce( ptorsion );
+
+  // Do periodic boundary conditions if necessary.
+  if (boxIn.HasBox()) {
+    //nonbond->setNonbondedMethod(OpenMM::NonbondedForce::CutoffPeriodic);
+    nonbond->setNonbondedMethod(OpenMM::NonbondedForce::PME);
+    mprintf("\tOpenMM cutoff (nm): %g\n", cut_);
+    nonbond->setCutoffDistance( cut_ );
+    Matrix_3x3 ucell = boxIn.UnitCell() * OpenMM::NmPerAngstrom;
+    ucell.Print("OpenMM unit cell");
+    system_->setDefaultPeriodicBoxVectors(
+      OpenMM::Vec3( ucell[0], ucell[1], ucell[2] ),
+      OpenMM::Vec3( ucell[3], ucell[4], ucell[5] ),
+      OpenMM::Vec3( ucell[6], ucell[7], ucell[8] ) );
+  }
+
+  // Add atoms to the system.
+  std::vector<int> oldToNew(topIn.Natom(), -1);
+  int newIdx = 0;
+  for (int idx = 0; idx != topIn.Natom(); idx++)
+  {
+    if (maskIn.AtomInCharMask(idx)) {
+      oldToNew[idx] = newIdx++;
+      system_->addParticle( topIn[idx].Mass() );
+      if (topIn.Nonbond().HasNonbond()) {
+        nonbond->addParticle(
+          topIn[idx].Charge(),
+          topIn.GetVDWradius(idx) * OpenMM::NmPerAngstrom * OpenMM::SigmaPerVdwRadius,
+          topIn.GetVDWdepth(idx) * OpenMM::KJPerKcal );
+      }
+    }
+  }
+  // Add bonds
+  std::vector< std::pair<int,int> >   bondPairs;
+  AddBonds(bondStretch, system_, bondPairs, topIn.Bonds(), topIn.BondParm(),  oldToNew, shakeHeavy_);
+  AddBonds(bondStretch, system_, bondPairs, topIn.BondsH(), topIn.BondParm(), oldToNew, shakeH_);
+
+  // Add angles
+  AddAngles(angleStretch, topIn.Angles(), topIn.AngleParm(), oldToNew);
+  AddAngles(angleStretch, topIn.AnglesH(), topIn.AngleParm(), oldToNew);
+
+  // Add dihedrals
+  AddDihedrals(ptorsion, topIn.Dihedrals(), topIn.DihedralParm(), oldToNew);
+  AddDihedrals(ptorsion, topIn.DihedralsH(), topIn.DihedralParm(), oldToNew);
+
+  // Populate nonbonded exclusions
+  nonbond->createExceptionsFromBonds(bondPairs, scaleEE_, scaleNB_);
```

Comment:
After this command you can go back and add exceptions for each 1-4 term individually in order to implement mixed 1-4 scaling. 

ParmEd has logic to do this in structure.py if you want to see a reference that has been validated. 

Just make sure you mark the exception to replace an existing one or you'll get an error.

Reply:
I actually wrote a C++-based Amber topology file parser + OpenMM system generator before I left for Lutron, and I transferred it to Todd Martinez so he could use it for something.

You can see the implementation here that obeys mixed 1-4 scaling in Amber topology files: https://github.com/tmartine68/omm_amber_cpp/blob/master/src/amberparm.cpp#L564-L584

Reply:
Nice, I didn't know about the exception stuff! I guess I should have tapped you to write the OpenMM support for CPPTRAJ! 😊